traits/src/storage.rs

```diff
@@ -0,0 +1,617 @@
+    fn delete_all_resumption_psk_secrets<GroupId: traits::GroupId<VERSION>>(
+        &self,
+        group_id: &GroupId,
+    ) -> Result<(), Self::Error>;
+
+    /// Deletes the own leaf index inside the group for the group with the given id.
+    fn delete_own_leaf_index<GroupId: traits::GroupId<VERSION>>(
+        &self,
+        group_id: &GroupId,
+    ) -> Result<(), Self::Error>;
+
+    /// Deletes any preference about whether to use the RatchetTreeExtension for the group with the given id.
+    fn delete_use_ratchet_tree_extension<GroupId: traits::GroupId<VERSION>>(
+        &self,
+        group_id: &GroupId,
+    ) -> Result<(), Self::Error>;
+
+    /// Deletes the GroupEpochSecrets for the group with the given id.
+    fn delete_group_epoch_secrets<GroupId: traits::GroupId<VERSION>>(
+        &self,
+        group_id: &GroupId,
+    ) -> Result<(), Self::Error>;
+
+    /// Clear the proposal queue for the grou pwith the given id.
+    fn clear_proposal_queue<GroupId: traits::GroupId<VERSION>>(
+        &self,
+        group_id: &GroupId,
+    ) -> Result<(), Self::Error>;
+
+    //
+    //    ---   deleters for crypto objects   ---
+    //
+
+    /// Delete a signature key pair based on its public key
+    ///
+    /// The signature key pair is not known to OpenMLS. This may be used by the
+    /// application
+    fn delete_signature_key_pair<SignaturePublicKey: traits::SignaturePublicKey<VERSION>>(
+        &self,
+        public_key: &SignaturePublicKey,
+    ) -> Result<(), Self::Error>;
+
+    /// Delete an encryption key pair for a public key.
+    ///
+    /// This is only be used for encryption key pairs that are generated for
+    /// update leaf nodes. All other encryption key pairs are stored as part
+    /// of the key package or the epoch encryption key pairs.
+    fn delete_encryption_key_pair<EncryptionKey: traits::EncryptionKey<VERSION>>(
+        &self,
+        public_key: &EncryptionKey,
+    ) -> Result<(), Self::Error>;
+
+    /// Delete a list of HPKE encryption key pairs for a given epoch.
+    /// This includes the private and public keys.
+    fn delete_encryption_epoch_key_pairs<
+        GroupId: traits::GroupId<VERSION>,
+        EpochKey: traits::EpochKey<VERSION>,
+    >(
+        &self,
+        group_id: &GroupId,
+        epoch: &EpochKey,
+        leaf_index: u32,
+    ) -> Result<(), Self::Error>;
+
+    /// Delete a key package based on the hash reference.
+    ///
+    /// This function only deletes the key package.
+    /// The corresponding encryption keys must be deleted separately.
+    fn delete_key_package<KeyPackageRef: traits::HashReference<VERSION>>(
+        &self,
+        hash_ref: &KeyPackageRef,
+    ) -> Result<(), Self::Error>;
+
+    /// Delete a PSK based on an identifier.
+    fn delete_psk<PskKey: traits::PskId<VERSION>>(
+        &self,
+        psk_id: &PskKey,
+    ) -> Result<(), Self::Error>;
+}
+
+// base traits for keys and values
```

Comment:
Let's add doc comments to all these.

Reply:
Do you mean the `Key` and `Entity` traits or all the individual ones? For the latter I'd suggest just adding a doc comment to the `traits` module that explains the idea behind the traits. What the reader needs to know is the concept, which is difficult to explain for individual items, and when you understand the concept, there is little value in a comment on an individual item.

Reply:
Yeah that works. We just have to have enough comments here so that people know what to do with it.